### PR TITLE
8263434: Dangling references after MethodComparator::methods_EMCP

### PR DESCRIPTION
**Pull request by shipilev**
SonarCloud reports the following problem in MethodComparator::methods_EMCP:
 "Address of stack memory associated with local variable 's_new' is still referred to by the global variable '_s_new' upon returning to the caller. This will be a dangling reference"

Code inspection reveals the assignment to static variables is only needed to pass them to helper methods. So, while this is not a detectable bug (yet), it is still cleaner not to expose stack variables in globals.

Additional testing:
 - [x] Linux x86_64 fastdebug `tier1`
 - [x] Linux x86_64 fastdebug, `vmTestbase_nsk_jvmti`

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8263434](https://bugs.openjdk.java.net/browse/JDK-8263434): Dangling references after MethodComparator::methods_EMCP 

### Reviewers
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to 9a43dca15c33d0b3b9d692e4f3cdc87654c716a6
 * [Serguei Spitsyn](https://openjdk.java.net/census#sspitsyn) (@sspitsyn - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2937/head:pull/2937`
`$ git checkout pull/2937`

### CODE REVIEW

File: src/hotspot/share/prims/methodComparator.cpp [hunk 1]
```diff
@@ -264,48 +259,48 @@ bool MethodComparator::args_same(Bytecodes::Code c_old, Bytecodes::Code c_new) {
   return true;
 }
 
-bool MethodComparator::pool_constants_same(int cpi_old, int cpi_new) {
-  constantTag tag_old = _old_cp->tag_at(cpi_old);
-  constantTag tag_new = _new_cp->tag_at(cpi_new);
+bool MethodComparator::pool_constants_same(int cpi_old, int cpi_new, ConstantPool* old_cp, ConstantPool* new_cp) {
```
Comment (coleenp) [inline src/hotspot/share/prims/methodComparator.cpp:262]: Can these be const?

File: src/hotspot/share/prims/methodComparator.cpp [hunk 2]
```diff
@@ -50,26 +45,26 @@ bool MethodComparator::methods_EMCP(Method* old_method, Method* new_method) {
...
+    if (!args_same(c_old, c_new, &s_old, &s_new, old_cp, new_cp))
       return false;
   }
   return true;
 }
 
-bool MethodComparator::args_same(Bytecodes::Code c_old, Bytecodes::Code c_new) {
+bool MethodComparator::args_same(Bytecodes::Code c_old, Bytecodes::Code c_new,
+                                 BytecodeStream* s_old, BytecodeStream* s_new,
+                                 ConstantPool* old_cp,  ConstantPool* new_cp) {
```
Comment (coleenp) [inline src/hotspot/share/prims/methodComparator.cpp:67]: Can these be const pointers too?

**Review by coleenp — APPROVED** [review]
This is much cleaner! Thank you, and thank you SonarCloud.

File: src/hotspot/share/prims/methodComparator.cpp (hunk 2 above)
Comment (shipilev) [reply src/hotspot/share/prims/methodComparator.cpp:67]: Yes, they can. Sprinkled.

File: src/hotspot/share/prims/methodComparator.cpp (hunk 1 above)
Comment (shipilev) [reply src/hotspot/share/prims/methodComparator.cpp]: Yes, they can. Sprinkled.

**Comment by dcubed-ojdk** [comment]
Please make sure that you get a review from someone on the Serviceability team.

**Review by sspitsyn — CHANGES_REQUESTED** [review]
Aleksey,
Sorry for being late to the party.
This looks good to me.
One nit: unneeded extra '()' what came from the original code:
`     if ((old_cp->klass_at_noresolve(cpi_old) != new_cp->klass_at_noresolve(cpi_new)))`

Thanks,
Serguei

**Comment by shipilev** [comment]
> One nit: unneeded extra '()' what came from the original code:
> ` if ((old_cp->klass_at_noresolve(cpi_old) != new_cp->klass_at_noresolve(cpi_new)))`

Sure, updated. Also updated copyrights.

**Review by sspitsyn — APPROVED** [review]

**Pull request closed without merging**